cirq-superstaq/cirq_superstaq/ops/qudit_gates.py

```diff
@@ -236,6 +236,130 @@ def __repr__(self) -> str:
         )
 
 
+class VirtualZPowGate(cirq.EigenGate):
+    """Applies a phase rotation between two successive energy levels of a qudit."""
+
+    def __init__(
+        self,
+        dimension: int = 2,
+        level: int = 1,
+        exponent: cirq.TParamVal = 1.0,
+        global_shift: float = 0.0,
+    ) -> None:
+        """Initializes as `VirtualZPowGate`.
+
+        Args:
+            dimension: The qudit dimension on which this gate will act.
+            level: The lowest energy level onto which this gate applies a phase. For example,
+                passing `level=2` a phase of `(-1)**exponent` will applied to energy levels
+                `[2, ..., dimension - 1]`. This is equivalent to phase shifting all subsequent
+                single-qudit gates acting in the `(1, 2)` subspace (assuming all other gates commute
+                with this one).
+            exponent: This gate's exponent (see `cirq.EigenGate` documentation for details).
+            global_shift: This gate's global phase (see `cirq.EigenGate` documentation for details).
+
+        Raises:
+            ValueError: If `dimension` is less than two.
+            ValueError: If `level` is invalid for the given dimension.
+        """
+        if dimension < 2:
+            raise ValueError("Invalid dimension (must be at least 2).")
+
+        # Allow e.g. level=-1 to specify the highest energy level
+        if not 0 < abs(level) < dimension:
+            raise ValueError(f"Invalid energy level for a dimension-{dimension} gate.")
+
+        self._dimension = dimension
+        self._level = level % dimension
+        super().__init__(exponent=exponent, global_shift=global_shift)
+
+    @property
+    def dimension(self) -> int:
+        """The qudit dimension on which this gate acts."""
+        return self._dimension
+
+    @property
+    def level(self) -> int:
+        """The lowest energy level onto which this gate applies a phase; for example if `level=2`
+        a phase of `(-1)**exponent` will applied to energy levels `[2, ..., dimension - 1]`. This
+        is equivalent to phase shifting all subsequent single-qudit gates acting in the `(1, 2)`
+        subspace (assuming all other gates commute with this one).
+        """
+        return self._level
+
+    def _qid_shape_(self) -> Tuple[int]:
+        return (self._dimension,)
+
+    def _with_exponent(self, exponent: cirq.TParamVal) -> VirtualZPowGate:
+        return VirtualZPowGate(
+            dimension=self._dimension,
+            level=self._level,
+            exponent=exponent,
+            global_shift=self._global_shift,
+        )
+
+    def _eigen_components(self) -> List[Tuple[float, npt.NDArray[np.float_]]]:
+        projector_phase = np.zeros(self._dimension)
+        projector_phase[self._level :] = 1
```

Comment:
`[self._level:]`?

Reply:
i thought so too but the formatter wants it this way :/